File: LeetCode/72.cpp
class Solution {
public:
    int minDistance(string word1, string word2) {
        int n1 = word1.size(), n2 = word2.size();
        std::vector<int> dp(n2 + 1);
        for (int j = 0; j < n2; j++) {
            dp[j + 1] = dp[j] + 1;
        }
        for (int i = 0; i < n1; i++) {
            int t = dp[0];
            ++dp[0];
            for (int j = 0; j < n2; j++) {
                int tt = dp[j + 1];
                if (word1[i] == word2[j]) {
                    dp[j + 1] = t;
                } else {
                    dp[j + 1] = std::min({t, dp[j], dp[j + 1]}) + 1;
                }
                t = tt;
            }
        }
        return dp[n2];
    }
};

File: LeetCode/72.py
class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        n1 = len(word1)
        n2 = len(word2)
        dp = [i for i in range(n2 + 1)]
        for i in range(n1):
            t = dp[0]
            dp[0] = i + 1
            for j in range(n2):
                tt = dp[j + 1]
                if word1[i] == word2[j]:
                    dp[j + 1] = t
                else:
                    dp[j + 1] = min(dp[j], dp[j + 1], t) + 1
                t = tt
        return dp[n2]
